plutus-wallet-api/src/Language/PlutusTx/StateMachine.hs

```diff
@@ -13,37 +15,65 @@
 module Language.PlutusTx.StateMachine(
       StateMachine(..)
     , StateMachineInstance (..)
+    , OldState(..)
+    , NewState(..)
+    , mkStateMachine
     , machineAddress
     , mkValidator
     ) where
 
-import           Language.PlutusTx.Prelude hiding (check)
-import qualified Language.PlutusTx as PlutusTx
+import qualified Language.PlutusTx                as PlutusTx
+import           Language.PlutusTx.Prelude        hiding (check)
+import           Ledger.Constraints
+import           Ledger.Constraints.OnChain       (checkPendingTx)
+import           Ledger.Constraints.TxConstraints (OutputConstraint (..))
 
-import           Ledger                    (Address)
+import           Ledger                           (Address, Value)
 import           Ledger.Typed.Scripts
-import           Ledger.Scripts (DataValue (..))
-import           Ledger.Validation         (PendingTx, PendingTxOut (..), PendingTxOutType (..), getContinuingOutputs, findData)
+import           Ledger.Validation                (PendingTx, PendingTx' (pendingTxItem),
+                                                   PendingTxIn' (pendingTxInValue))
+import           Ledger.Value                     (isZero)
+
+data OldState s = OldState { oldData :: s, oldValue :: Value }
+data NewState s = NewState { newData :: s, newValue :: Value, newConstraints :: [TxConstraint] }
```

Comment:
Yeah I think you're right, it's better to separate out the constraints. I wanted to exclude constraints that relate to typed inputs and outputs from the constraints emitted by the state machine (because those kind of constraints are given by the transition function itself), so I used `[TxConstraint]`, but the same effect is achieved by `TxConstraints Void Void`, which I'm going to change it to.